Improved usage of the client library

mattheworiordan opened the pull request:
@stannedelchev, I went to review the code, and to be honest there was just too much to comment on.  The code is good, I like what you have done, and it was very easy for me to work with this.  However, you were using the client library in ways that were unnecessary - for example, leaving presence, closing the channel and then closing the connection is unnecessary as if you close a connection, all other state is lost too so the user immediately leaves the connection.

Please don't look at the complete set of changes as it won't make any sense.  Please can you look through all of the commits as I expect you will then understand why I made the changes, and importantly, will then know how to do the same with the iOS, Android and Xamarin versions.

When you want, happy to chat this through.

Note: I found a few small bugs that are preventing this working perfectly, connection recovery (not resume from disconnected) is failing, and `untilAttach` that is used for history requests to retrieve all messages up until the channel was attached, is not working.  It's an easy fix, so I will get that looked at tomorrow, and we can then remove this temporary code.

I have pushed this to http://ably-phonegap-cordova-demo.herokuapp.com/

FYI @SimonWoolf and @paddybyers

stannedelchev commented:
@mattheworiordan thanks for the fixes. I'm glad you liked the code as I'm not really proficient in web technologies.
I like the new restructuring, `.once()` and proper Ably event subscription really cleaned up the mess of callbacks.

Some notes:
- I fixed a minor typo in `leftTrim()`'s usage, that prevented clicks on usernames to work properly
- Added some missing semicolons that WebStorm pointed out
- Mobile onResume() functionality is broken: on resuming the app, [the screen just clears](http://i.imgur.com/1fpyNR3.png) and the user is stuck. Sending a message only shows the "sending" notification, but that's all. I'll debug this and let you know what I've found.
- I really like the new styling, we just need to fix it for mobile devices as well. Scrolling to the newest element works fine on browser, but on mobile the input pane is taller (IMO a bug) and it may be better to use something like `$(...).height()` instead of `100` in the scrolling calculations.
- Reverted the `<meta>` CSP as it is [recommended by the Cordova Whitelist Plugin](https://github.com/apache/cordova-plugin-whitelist#content-security-policy) when running the hybrid app on a mobile device. 
  If the plugin is missing, no connection can be made to Ably and the app is broken. If the plugin is installed, but the `<meta>` not added, there is an error and periodic warnings in the Dev Console, which IMO doesn't make for a nice dev experience, although the app works fine. 

I'm fine with creating a less prohibitive CSP, or removing it altogether, if you think the dev console warnings are fine while debugging the mobile app.
 What errors were you referring to when you removed the `<meta>` tag?

:+1: for merging.

mattheworiordan commented:
@stannedelchev, re: mobile resume, I knew about that, but didn't worry. But it needs to work for now, so I have implemented a workaround in ec5c2a5 until https://github.com/ably/ably-js/issues/95 is fixed.

I have tried a less restrictive CSP now, can you try on Cordova and see if it still shows warnings.  In Chrome, I get lots of errors, see 8443208 for detail.

I have added three commits, if happy and once CSP is resolved, please go ahead and merge :+1:
